File: code/ex472.java
// Trie, DFS

// Define trieNode data structure
class trieNode {
    char val;
    trieNode[] children;
    boolean ifEnd;
    
    // Init
    public trieNode(char _val) {
        val = _val;
        children = new trieNode[26];
        ifEnd = false;
    }
}

class Trie {
    trieNode root;
    
    /** Initialize your data structure here. */
    public Trie() {
        root = new trieNode(' ');
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        trieNode p = root;       
        for (int i = 0; i < word.length(); i++) {
            if (p.children[word.charAt(i) - 'a'] == null) {
                p.children[word.charAt(i) - 'a'] = new trieNode(word.charAt(i));
                p = p.children[word.charAt(i) - 'a'];
            } else {
                p = p.children[word.charAt(i) - 'a'];
            }
        }    
        p.ifEnd = true;
    }
    
    /** Direct to the end of the string in the trie tree*/
    public trieNode direct(String word) {
        trieNode p = root;
        for (int i = 0; i < word.length(); i++) {
            if (p.children[word.charAt(i) - 'a'] != null)
                p = p.children[word.charAt(i) - 'a'];
            else    return null;
        } 
        return p;
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        trieNode p = direct(word);      
        if (p != null && p.ifEnd == true)    return true;
        else    return false;
    }
}

class Solution {
    public List<String> findAllConcatenatedWordsInADict(String[] words) {
        Trie trie = new Trie();
        
        for (String s : words)
            trie.insert(s);
        
        List<String> res = new ArrayList<>();
        for (String s : words) {
            for (int i = 0; i < s.length(); i++) {
                if (trie.search(s.substring(0, i + 1)) && dfs(s, trie, i + 1)) {
                    res.add(s);
                    break;
                }
            }   
        }
 
        return res;
    }
    
    private boolean dfs(String s, Trie trie, int start) {
        boolean res = false;
        for (int i = start; i < s.length(); i++) {
            if (trie.search(s.substring(start, i + 1)) && i + 1 == s.length())
                res = true;
            else if (trie.search(s.substring(start, i + 1)))    
                res = dfs(s, trie, i + 1);
            if (res)    return true;
        }
        return res;
    }
}

File: code/ex472.cpp
struct Trie {
    // Definition of TrieNode
    struct TrieNode {
        vector<TrieNode*> children;
        bool ifEnd;
        TrieNode() {
            children = vector<TrieNode*>(26);
            ifEnd = 0;
        }
    };

    //////////////////////////////////////////////////////////
    TrieNode* root;
    
    Trie() {
        root = new TrieNode();
    }
    
    // Insert a string
    void insert(string s) {
        TrieNode* p = root;
        for (char c : s) {
            if (p->children[c - 'a'] == NULL)
                p->children[c - 'a'] = new TrieNode();
            p = p->children[c - 'a'];
        }
        p->ifEnd = 1;
    }
    
    // Search for a string
    bool search(string s) {
        TrieNode* p = root;
        for (char c : s) {
            if (p->children[c - 'a'] == NULL)    return 0;
            p = p->children[c - 'a'];
        }
        return p->ifEnd;
    }
};


class Solution {
    Trie* trie = new Trie();
    
public:
    vector<string> findAllConcatenatedWordsInADict(vector<string>& words) {
        for (string word : words)    trie->insert(word);
        
        vector<string> res;
        unordered_map<string, bool> memo;
        for (string word : words) {
            int n = word.length();
            for (int i = 1; i <= n - 1; i++) {
                string first = word.substr(0, i), second = word.substr(i);
                if (trie->search(first) && backtracking(second, memo)) {
                    res.push_back(word);
                    break;
                }
            }
        }
        
        return res;
    }
    
private:
    bool backtracking(string word, unordered_map<string, bool>& memo) {
        if (word.length() == 0)    return 1;
        if (memo.count(word) > 0)    return memo[word];
        bool res = 0;
        for (int i = 1; i <= word.length(); i++) {
            string cur = word.substr(0, i), nxt = word.substr(i);
            if (trie->search(cur) && backtracking(nxt, memo)) {
                res = 1;
                break;
            }
        }
        memo[word] = res;
        return res;
    }
};